ManageIQ/manageiq-performance
# Optionally add stackprof to the installation script generation

Purpose
-------
On some systems, it will be requested to do extra profiling to the appliance, and this allows the installation script to include and install stackprof onto the appliance to be used with `manageiq-performance`.


Implementation
--------------
A few things were added to make this work

**`GemBase64::find_gemspec_for`:**  This method will find the latest spec for a given gem, either one that exists, or downloading one that is already available on the system.  Being that stackprof is an optional dependency for this project, it may or may not exist on the developers machine at the time of generation (even though it is currently a development dependency, it is possible that might change in the future, or this script is run without bundling).

**`GemBase64::gem_as_base64_string` and `GemBase64::gem_as_tar_io` updates:**  These two methods now take optional arguments to pass in either their own gemfile contents or gemspec respectively.  This allows the contents from them to be reused for stackprof.

**Updates to the template:**  The template now has multiple conditionals in it around adding stackprof content.  This is activated when the `@stackprof_gemspec` and `@stackprof_gem_base64_string` instance variables are present when rendering the template.  These are added by the new task...

**Added new task `include_stackprof`:**  This task is in charge of adding the instance variables `@stackprof_gemspec` and `@stackprof_gem_base64_string` so the template can make use of it.  Makes use of the new features in `GemBase64`.


TODO
----
* [x] Clean up test output


QA
--
* Generate a script using `$ rake include_stackprof generate_install_script`
* Run that script on an appliance, and confirm that it works, and the server can be restarted
* Hit a few routes and confirm that stackprof data is generated along with the other middleware plugins
